core/indigo-core/common/math/algebra.h

```diff
@@ -822,5 +824,234 @@ namespace indigo
+                return false;
+        return true;
+    }
+
+    inline std::vector<Vec2f> getPointsInsideConvexPolygon(const std::vector<Vec2f>& polygon1, const std::vector<Vec2f>& polygon2)
+    {
+        std::vector<Vec2f> result;
+        result.reserve(polygon1.size());
+        std::copy_if(polygon1.begin(), polygon1.end(), std::back_inserter(result), [&](auto& p) { return isPointInConvexPolygon(p, polygon2); });
+        return result;
+    }
+
+    inline float convexPolygonArea(const std::vector<Vec2f>& poly)
+    {
+        float area = 0.0f;
+        size_t n = poly.size();
+        for (size_t i = 0; i < n; ++i)
+        {
+            const Vec2f& p1 = poly[i];
+            const Vec2f& p2 = poly[(i + 1) % n];
+            area += Vec2f::cross(p1, p2);
+        }
+        return std::abs(area) * 0.5f;
+    }
+
+    inline bool isInside(const Vec2f& edgeStart, const Vec2f& edgeEnd, const Vec2f& point)
+    {
+        return (edgeEnd.x - edgeStart.x) * (point.y - edgeStart.y) - (edgeEnd.y - edgeStart.y) * (point.x - edgeStart.x) <= 0;
+    }
+
+    inline bool convexPolygonsIntersect(const std::vector<Vec2f>& poly1, const std::vector<Vec2f>& poly2)
+    {
+        auto project = [](const std::vector<Vec2f>& poly, const Vec2f& axis) {
+            auto [min, max] = std::minmax_element(poly.begin(), poly.end(), [&](const Vec2f& a, const Vec2f& b) { return a * axis < b * axis; });
+            return std::pair{*min * axis, *max * axis};
+        };
+
+        auto overlap = [](const auto& range1, const auto& range2) { return !(range1.first > range2.second || range2.first > range1.second); };
+
+        auto getNormals = [](const std::vector<Vec2f>& poly) {
+            std::vector<Vec2f> normals;
+            for (size_t i = 0; i < poly.size(); ++i)
+            {
+                Vec2f edge = poly[(i + 1) % poly.size()] - poly[i];
+                normals.emplace_back(-edge.y, edge.x);
+            }
+            return normals;
+        };
+
+        auto normals1 = getNormals(poly1), normals2 = getNormals(poly2);
+        for (const auto& normal : normals1)
+            if (!overlap(project(poly1, normal), project(poly2, normal)))
+                return false;
+        for (const auto& normal : normals2)
+            if (!overlap(project(poly1, normal), project(poly2, normal)))
+                return false;
+        return true;
+    }
+
+    inline Vec2f computeIntersection(const Vec2f& s, const Vec2f& e, const Vec2f& edgeStart, const Vec2f& edgeEnd)
+    {
+        float dx1 = e.x - s.x, dy1 = e.y - s.y, dx2 = edgeEnd.x - edgeStart.x, dy2 = edgeEnd.y - edgeStart.y;
+        float det = dx1 * dy2 - dy1 * dx2;
+        if (std::abs(det) < 1e-6)
+            return s;
+        float t = ((edgeStart.x - s.x) * dy2 - (edgeStart.y - s.y) * dx2) / det;
+        return {s.x + t * dx1, s.y + t * dy1};
+    }
+
+    inline std::vector<Vec2f> convexClip(const std::vector<Vec2f>& subject, const std::vector<Vec2f>& clip)
```

Comment:
Please add comment that this is "Sutherland–Hodgman algorithm"

Reply:
fixed